ruby/computer_science/time_complexity.md

```diff
@@ -0,0 +1,306 @@
+
+- `O(1)` - Constant Complexity
+- `O(log N)` - Logarithmic Complexity
+- `O(N)` - Linear Complexity
+- `O(N log N)` - N x log N Complexity
+- `O(&#178;)` - Quadratic Complexity
+- `O(&#179;)` - Cubic Complexity
+- `O(2&#8319;)` - Exponential Complexity
+- `O(N!)` - Factorial Complexity
+
+#### O(1) - Constant Complexity
+
+To understand Constant Complexity let's use a simple array.
+
+~~~ruby
+arr = [1, 2, 3, 4, 5]
+~~~
+
+If we want to look up what is at index 2, we can get to the element using `arr[2]` which would give us back `3`. This takes just one step. If we double our array...
+
+~~~ruby
+arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+~~~
+
+We can still access any element in just one step. `arr[7]` gives me `8` in a single step. Our array can keep growing and we can always access any element in a single step. It's constant. Hence we have `O(1)`.
+
+Looking up something in one step is as good as it gets for time complexity.
+
+While we're looking at the simplest form of Big O, let's take a look at one of its little gotchas to keep in mind. You may have thought a moment ago, is it really just one step? The answer is technically no, in reality the computer must first look up where the array is in memory, then from the first element in the array it needs to jump to the index argument provided. That's at least a couple of steps. So you wouldn't be wrong for writing something like `O(1 + 2(steps))`. However, the 2 steps are merely incidental. With an array of 10,000 elements it still takes the same amount of steps as if the array was 2 elements. Because of this, Big O doesn't concern itself with these incidental numbers. They don't provide any context to how the complexity grows when the data size changes, because they are constant, and so in Big O they are dropped. Big O only wants to tell us an algorithm's complexity relative to the size of the input.
+
+Do the number of steps matter? Yes, they might. We'll touch on when this may be the case a little later.
+
+#### O(log N) - Logarithmic Complexity
+
+Logarithmic Complexity tells us that the numbers of steps an algorithm takes increases by 1 as the data doubles. That's still pretty efficient when you think about it. Going from 5,000 to 10,000 data elements and only taking one additional step can scale really well.
+
+One such algorithm that does this is Binary Search. It only works on sorted arrays but if you have an array of 10 items in sorted order and wanted to know if it had the number `7`, Binary Search would guess the middle item of the array and see what is there. Because the array is sorted, if the number at the middle index was `5`, then we know anything to the left of that index cannot be the number 7, as those items must be lower than 5 in a sorted array. Therefore in just one step we've eliminated half of the array. We can do the same with the remaining half. We can guess the middle index and see if it's 7. Half of that (half of an array) array eliminated again.
```

Comment:
We can add code representation of the array here, so learners can easily reference the array with the text explanation. Something like this:
```suggestion
One such algorithm that does this is Binary Search. It only works on sorted arrays but if you have an array of 10 items in sorted order

~~~ruby
arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
~~~

and wanted to know if it had the number `7`, Binary Search would guess the middle item of the array and see what is there. Because the array is sorted, if the number at the middle index was `5`, then we know anything to the left of that index cannot be the number 7, as those items must be lower than 5 in a sorted array. Therefore in just one step we've eliminated half of the array. We can do the same with the remaining half. We can guess the middle index and see if it's 7. Half of that (half of an array) array eliminated again.
```